src/compiler/crystal/semantic/type_guess_visitor.cr

```diff
@@ -631,6 +631,14 @@ module Crystal
       return type if type
 
       type = guess_type_call_with_type_annotation(node)
+
+      # If the type is unbound (uninstantiated generic) but the call
+      # wasn't something like `Gen(Int32).something` then we can never
+      # guess a type, the type is probably inferred from type restrictions
+      if !obj.is_a?(Generic) && type.try &.unbound?
```

Comment:
To be honest, I'm not super happy with this solution but:
- it kind of makes sense
- it gets rid of a bug that seems to be very popular (it was reported countless times already)
- fixing it another way might involve a bigger refactor but I don't have time for it right now